### PR TITLE
Make config file formats flexible and dog-frendly

### PR DESCRIPTION
**Pull request by frozencemetery**
Right now, rpminspect is hardwired to YAML for configuration.  While it's probably not possible to migrate off wholesale right now, we can provide alternatives.  Introduce a new parsing interface (parser.h), and implement it for [YAML](https://yaml.org/), [DSON](https://dogeon.xyz/), and [JSON](https://www.json.org/json-en.html), thereby allowing configuration to be in any of those formats as well as providing an abstraction over parsing.

So, for example, adding [TOML](https://toml.io/) support just requires implementing parser.h for TOML.  (If you actually want TOML support, [tomlc99](https://github.com/cktan/tomlc99) looks okay for the heavy lifting, but I'm happy to write out a TOML parser if you want one.  That sounds like a good time.)

Coping with YAML is hard enough, but libyaml makes it even harder by shifting the burden of parsing onto the application itself.  As a result, there are a few correctness issues with rpminspect yaml files today (and some that aren't strictly incorrect, but are rather odd and would be nice to have fixed since compat is technically broken anyhow).

Move the burden of parsing to [libtyranny](https://github.com/frozencemetery/libtyranny), which builds on libyaml but provides a more friendly tree structure (and performs the missing validity checking).  Even vendoring a copy of libtyranny, this is shorter and easier to reason than using libyaml directly.  Assuming the project wants this, I'm happy to discuss how to handle this dependency "properly" - and just for variety, I handled the cdson dependency a different way (submodule present for fallback).  A third way could be to fold the libtyranny code into parse_yaml.c completely.

There's some intermediate stuff (or maybe even final stuff) here you may not want.  I'm happy to split/recombine commits here, or make separate PRs for chunks of this if that's helpful.

For anyone reviewing not familiar with parsing YAML, some (mostly true) context:

- YAML supports both quoted and bare (i.e., unquoted) strings.  While the spec   defines other types, actually using them in an intermediate representation is therefore [a trap](https://hitchdev.com/strictyaml/why/implicit-typing-removed/).  I recommend treating string as the only base type.

- Well, string and NULL, as dictionary values can be NULL (i.e., when the colon is present but nothing is on the other side, like `foo:`).

- Speaking of dictionaries, YAML calls these "mappings", while arrays are "sequences".

- There are two ways to write YAML: the whitespace-y indentation-matters "normal" way, which is called "flow", and the looks-like-json way, which is called "block".  They can be mixed to some degree.

- YAML is a superset of JSON.  We could use the YAML parser to parse JSON files too, instead of linking in json-c.  This is not my preference, as it effectively migrates everything *to* YAML and I want to migrate away.

- YAML supports some meta-file options as well.  These include versioning, aliases, tags, and anchors.

- rpminspect before these changes did not support any of them.  After these changes, it doesn't either.  libtyranny doesn't because not once have I seen them used in the wild, let alone used for anything good.

- The "---" that appears in some YAML files separates header meta-file stuff from the main file.  It is optional when there is no such stuff.

- YAML supports python-style line comments with '#'.  Strictly speaking, JSON and DSON have no comment support.  Parsers ignore comments.

### CODE REVIEW

**Comment by dcantrell** [comment]
Well this is a welcome and nice surprise.  This is a stretch goal I had for rpminspect in 2023.  I absolutely regret my decision to go with YAML for configuration files.  Initially I was using iniparser or some C .INI-style parsing thing, but since there's no INI-standard for lists I quickly outgrew that and didn't want to create a new config file format.  YAML had inertia on my team and there was a push to use that for config files for our tools because the argument was it would be easy to integrate with other tools using YAML, like Ansible.  In the end I didn't really care except that I didn't want to be in the business of defining a new config file.  Ironically I ended up doing just that.

One of the big downsides is that YAML is in the wild, so support for that has to remain for the long term.  I am ok with that and can mark it as deprecated in a future release with plans to stop supporting it.  But for now rpminspect needs to support not just YAML but the janky way I implemented some of the configuration file settings in rpminspect.yaml.  I do not want to break this as it was hard enough to get people to write these files in the first place and put them in dist-git, to break syntax now would not be good.

I really like this patch set, but I want to go through it in more detail.  Right now I don't see a reason to break up commits or anything, but I will have more on an opinion when the work year starts.

The first things I've noticed as I've glanced through this that I would prefer an alternative way or at least discuss options:

- I'm not a fan of submodules.  It may not be avoidable in some cases, but I would prefer to avoid them if at all possible.  I'd like to see cdson and libtyranny added to Fedora and EPEL so they are available for Koji builds.  For CI purposes, I don't mind cloning and installing build requirements in the container.  I do that for a lot of things already.
- I know I already have a copy of libxdiff in this project, but that one is weird and there's no real way around it.  I also have uthash.h which there is no way around because that's how the maintainer distributes it.  We do have a package for it in Fedora which is kind of cute, but I can carry a header file.
- Adding TOML support would be nice.  I have looked at tomlc99 briefly, but nothing beyond that.  I don't mind that being a separate PR entirely.
- I see that the generic.yaml file changed a number of syntax aspects.  I want to make sure the existing style still works.  A number of package maintainers make use of the badfuncs section to redefine things.  I assume it does because the test suite passes, but I'd like to look in more detail.
- I noticed a lot of changes to the license inspect, obviously, but I want to make sure the matching works as it has been.  It's unusual because the license inspection matches the largest substring but also does this sort of half-assed tokenization because how we defined things for spec files was not parse-friendly.  The usual example I give is Perl packages that have "GPL+ or Artistic" and in the license database we have GPL+ as approved, Artistic as not approved, but the entire substring "GPL+ or Artistic" as approved.  The substring was added likely to work around the problem introduced in the Perl package metadata.  If the license inspection breaks it down to GPL+ and Artistic tokens and checks those individually, then it will fail because Artistic is not approved.  This is also going away with the move to SPDX, but what we have now is live and in the wild so we need to support it.
- Coding style preferences.  I wrap comments at 80 characters but I let code go longer.  For functions, I put { on its own line.  Blank lines before and after blocks.  And I am particular about my switch statements.  I can comment on these individually or make changes later, it doesn't really matter to me.  But I'd like stuff consistent.

Everything about this patch set is great and moving away from my terrible YAML parsing loop is great.  I hate that function I wrote and any time I had to expand the config file it would take days to sort out and make sure I didn't break for other things.

I will continue reviewing later this week.

**Comment by frozencemetery** [comment]
> But for now rpminspect needs to support not just YAML but the janky way I implemented some of the configuration file settings in rpminspect.yaml. I do not want to break this as it was hard enough to get people to write these files in the first place and put them in dist-git, to break syntax now would not be good.
>
> ...
>
> I see that the generic.yaml file changed a number of syntax aspects. I want to make sure the existing style still works. A number of package maintainers make use of the badfuncs section to redefine things. I assume it does because the test suite passes, but I'd like to look in more detail.

Understood.  Most backward compat can be added, but badfuncs and javabytecode are trouble:

```yaml
badfuncs:
    - abadfunc
    - anotherbadfunc
    allowed:
        /usr/path/whatever/*: 
              - abadfunc
```

For compat, we can handle the sequence-only case (when allowed isn't present), and we can handle the mapping-only case (when only allowed is present).  Quite possibly no one has ever defined "allowed", as pyyaml won't work with it (and, therefore, neither will the test suite).  Do you think that would be sufficient?

> I'd like to see cdson and libtyranny added to Fedora and EPEL so they are available for Koji builds.

I did make a [COPR for cdson](https://copr.fedorainfracloud.org/coprs/rharwood/cdson/) and would be fine with taking that into the main distro.  However, I do not want to maintain libtyranny long-term; I'm therefore hesitant to go down that road without more of a plan.  (I'm also fine with folding it in entirely rather than vendoring.)

> Adding TOML support would be nice. I have looked at tomlc99 briefly, but nothing beyond that. I don't mind that being a separate PR entirely.

This'll actually run into the same problem: there aren't TOML parsers in Fedora that we could use, as far as I can tell, so someone would need to package and own tomlc99.  (For transparency, it's the only one in C listed [on the official toml-lang wiki](https://github.com/toml-lang/toml/wiki#implementations), while the C++ ones listed there are explicitly C++-only; the only C one present in Fedora doesn't support 1.0.0.)

> I noticed a lot of changes to the license inspect, obviously, but I want to make sure the matching works as it has been. It's unusual because the license inspection matches the largest substring but also does this sort of half-assed tokenization because how we defined things for spec files was not parse-friendly.

I didn't touch the tokenization, but definitely better to be safe.  It did take me a bit to figure out what was going on, as the inspection uses parser objects as contexts directly.  So messing with the parser interfaces messes with the context objects...

Another thing to consider is that while the parser interface works for JSON, the point was more to show that it's general than actually rip out json-c calls.  (I think json-c is okay, albeit overengineered.)  Which is to say: if the licensedb format isn't likely to change, and you're not interested in having the main config file be JSON, I wouldn't be offended if you pass on all of json_parser.  (Same goes for dson_parser.)

> Coding style preferences. I wrap comments at 80 characters but I let code go longer. For functions, I put { on its own line. Blank lines before and after blocks. And I am particular about my switch statements. I can comment on these individually or make changes later, it doesn't really matter to me. But I'd like stuff consistent.

Added more blank lines around blocks in self-review; hopefully that helps.  Am afraid you're going to have to point out the rest - in particular, I don't know how to "let code go longer" (my brain doesn't work that way) so I'm endeavoring to wrap non-strings at 80 characters.  libtyranny isn't trying to meet your style right now, but if you want it to live in-tree I can certainly change that.

**Comment by frozencemetery** [comment]
Rebased onto main.  Feel free to cherry-pick from this; I'm also happy to open a separate PR for the noncontroversial bugixes.

**Comment by dcantrell** [comment]
> Thanks for the review!
> 
> > But for now rpminspect needs to support not just YAML but the janky way I implemented some of the configuration file settings in rpminspect.yaml. I do not want to break this as it was hard enough to get people to write these files in the first place and put them in dist-git, to break syntax now would not be good.
> > ...
> > I see that the generic.yaml file changed a number of syntax aspects. I want to make sure the existing style still works. A number of package maintainers make use of the badfuncs section to redefine things. I assume it does because the test suite passes, but I'd like to look in more detail.
> 
> Understood. Most backward compat can be added, but badfuncs and javabytecode are trouble:
> 
> ```yaml
> badfuncs:
>     - abadfunc
>     - anotherbadfunc
>     allowed:
>         /usr/path/whatever/*: 
>               - abadfunc
> ```
> 
> For compat, we can handle the sequence-only case (when allowed isn't present), and we can handle the mapping-only case (when only allowed is present). Quite possibly no one has ever defined "allowed", as pyyaml won't work with it (and, therefore, neither will the test suite). Do you think that would be sufficient?

Probably so.  Right now all people do (that I have seen) is override the badfuncs list, so the ability to handle a sequence-only case should handle those.

> > I'd like to see cdson and libtyranny added to Fedora and EPEL so they are available for Koji builds.
> 
> I did make a [COPR for cdson](https://copr.fedorainfracloud.org/coprs/rharwood/cdson/) and would be fine with taking that into the main distro. However, I do not want to maintain libtyranny long-term; I'm therefore hesitant to go down that road without more of a plan. (I'm also fine with folding it in entirely rather than vendoring.)

I understand that.  Given those options, I would prefer folding in libtyranny rather than using submodules.  Maybe this is something worth trying to get in to libyaml (later).

cdson as a package would be nice.  If you're going to do that, I'll need that to happen before I can release v1.11 (which I've said I've been trying to do for two months now, but I really do mean it).  But also v1.11 can't happen until I merge this patch set.

> > Adding TOML support would be nice. I have looked at tomlc99 briefly, but nothing beyond that. I don't mind that being a separate PR entirely.
> 
> This'll actually run into the same problem: there aren't TOML parsers in Fedora that we could use, as far as I can tell, so someone would need to package and own tomlc99. (For transparency, it's the only one in C listed [on the official toml-lang wiki](https://github.com/toml-lang/toml/wiki#implementations), while the C++ ones listed there are explicitly C++-only; the only C one present in Fedora doesn't support 1.0.0.)

Eh, forget it then.  I thought it came along for free for some reason because I cannot remember what format is a descendant of another.  If someone wants TOML (they will), patches welcome.

> > I noticed a lot of changes to the license inspect, obviously, but I want to make sure the matching works as it has been. It's unusual because the license inspection matches the largest substring but also does this sort of half-assed tokenization because how we defined things for spec files was not parse-friendly.
> 
> I didn't touch the tokenization, but definitely better to be safe. It did take me a bit to figure out what was going on, as the inspection uses parser objects as contexts directly. So messing with the parser interfaces messes with the context objects...

The work we've been doing in fedora-license-data has taken care of a lot of this bad data in the license database, so I am really only partially worried about the compatibility here.  When I get to that part in the code, I can check in more detail.

> Another thing to consider is that while the parser interface works for JSON, the point was more to show that it's general than actually rip out json-c calls. (I think json-c is okay, albeit overengineered.) Which is to say: if the licensedb format isn't likely to change, and you're not interested in having the main config file be JSON, I wouldn't be offended if you pass on all of json_parser. (Same goes for dson_parser.)

The format isn't going to change, though the fedora-license-data project may start publishing in other formats.  But since the JSON format exists, I'm just using that.  I would prefer to remove the json-c dependency, so I'd rather have this work incorporated.  My choice of json-c was that it's in C and I could mostly understand the API.

> > Coding style preferences. I wrap comments at 80 characters but I let code go longer. For functions, I put { on its own line. Blank lines before and after blocks. And I am particular about my switch statements. I can comment on these individually or make changes later, it doesn't really matter to me. But I'd like stuff consistent.
> 
> Added more blank lines around blocks in self-review; hopefully that helps. Am afraid you're going to have to point out the rest - in particular, I don't know how to "let code go longer" (my brain doesn't work that way) so I'm endeavoring to wrap non-strings at 80 characters. libtyranny isn't trying to meet your style right now, but if you want it to live in-tree I can certainly change that.

I will comment on individual lines.  What I meant by "let code go longer" is that unless a line is a comment, I let it extend past 80 columns.  I wrap comment blocks at 80 columns.  I know that seems horribly inconsistent, but it is consistent with the rest of my horribly inconsistent code.

I naturally make exceptions, the biggest being large if expressions.  I will add individual comments for what I would like to see.  I would prefer libtyranny matching up the style, but that one is ok to me.  I did not do that for libxdiff.

Comments coming on the patches.

File: meson.build [hunk 1]
```diff
@@ -264,6 +264,10 @@ endif
 
 m = declare_dependency(link_args : ['-lm'])
 
+# cdson, with in-tree fallback
+cdson = dependency('cdson', fallback : ['cdson', 'cdson_dep'],
+		   default_options: ['default_library=static'])
```
Comment (dcantrell) [inline meson.build:269]: This should change with cdson becoming a real package.

File: test/baseclass.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
 #
 
+import dson
```
Comment (dcantrell) [inline test/baseclass.py:6]: So here's where it's probably annoying.  With multiple formats for config files, I want the test suite to test all possibilities.  I'd probably just make new environment variables that the Makefile can use.  Like RPMINSPECT_YAML, RPMINSPECT_DSON, and RPMINSPECT_JSON.  Then baseclass and other can figure out what to do from that.

File: data/generic.yaml [hunk 3]
```diff
@@ -597,11 +597,8 @@ annocheck:
     #
     # This section is optional.  If no annocheck tests are defined
     # here, rpminspect will skip the annocheck inspection.
-    #
-    # These job entries may be listed here or under a block named
-    # 'jobs:' if you prefer.
     jobs:
-        - hardened: --ignore-unknown
+        hardened: --ignore-unknown
```
[the remark this answers is not in the document]
Comment (frozencemetery) [reply data/generic.yaml:601]: tl;dr: it's the difference between a dict and a list of dicts.

`foo: bar` is a dict (YAML calls it a "mapping").  The JSON equivalent would be `{ "foo": "bar" }`.

In JSON, we write arrays like `["foo", "bar"]`.  The YAML equivalent of this (called a "sequence", here in flow) would be:

```yaml
- foo
- bar
```

In other words, the leading dashes indicate that it's an array, and each line is an element of the array.  Without the dash, it's not an array but a dictionary:

```yaml
foo: bar
baz: buzz
```

is equivalent to the JSON

```json
{
    "foo": "bar",
    "baz": "buzz"
}
```

Having the dashes turns it into an array, producing an array of dictionaries instead of a single unified dictionary:

```yaml
- foo: bar
- baz: buzz
```

is the JSON

```json
[
    { "foo": "bar" },
    { "baz": "buzz" }
]
```

which has an extra level of indirection.

File: include/tyranny.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
+/* SPDX-License-Identifier: MPL-2.0 */
+/* Copyright the libtyranny contributors */
+
+/* This file was originally part of libtyranny:
+ * https://github.com/frozencemetery/libtyranny/ */
+
```
[the remark this answers is not in the document]
Comment (frozencemetery) [reply include/tyranny.h:6]: I would disappear it into parse_yaml.c unless you wanted it to persist.

File: lib/builds.c [hunk 5]
```diff
@@ -341,67 +346,25 @@ static int download_build(struct rpminspect *ri, const struct koji_build *build)
...
-                    if (token.type != YAML_STREAM_END_TOKEN) {
-                        yaml_token_delete(&token);
-                    }
-                } while (token.type != YAML_STREAM_END_TOKEN);
-
-                /* destroy the YAML parser, close the input file */
-                yaml_parser_delete(&parser);
-
-                if (fclose(fp) != 0) {
-                    err(RI_PROGRAM_ERROR, "fclose");
```
[the remark this answers is not in the document]
Comment (frozencemetery) [reply lib/builds.c:403]: f

File: lib/init.c [hunk 6]
```diff
@@ -44,106 +43,6 @@ static const char *BIN_PATHS[] = {"/bin", "/sbin", "/usr/bin", "/usr/sbin", NULL
...
-    BLOCK_UPSTREAM = 70,
-    BLOCK_VIRUS = 71,
-    BLOCK_ENVIRONMENT = 72,
-    BLOCK_LICENSEDB = 73,
-    BLOCK_DEBUGINFO = 74,
-    BLOCK_PATCH_AUTOMACROS = 75,
-    BLOCK_MODULARITY = 76,
-    BLOCK_ANNOCHECK_PROFILE = 77
-};
-
```
[the remark this answers is not in the document]
Comment (frozencemetery) [reply lib/init.c:146]: Hopefully this results in less and not more!

File: meson.build [hunk 7]
```diff
@@ -7,7 +7,7 @@ project('rpminspect',
             'werror=true',
             'buildtype=debugoptimized'
         ],
-        license : 'GPL-3.0-or-later AND LGPL-3.0-or-later AND Apache-2.0 AND MIT')
+        license : 'GPL-3.0-or-later AND LGPL-3.0-or-later AND Apache-2.0 AND MIT AND MPL-2.0')
```
[the remark this answers is not in the document]
Comment (frozencemetery) [reply meson.build:10]: If we fold libtyranny in, I'm fine to just relicense it like a normal rpminspect file.

File: include/tyranny.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (frozencemetery) [reply include/tyranny.h:6]: Amending that: since in project style structs don't live in .c files, a header file needs to exist for them, so reduced version of this persists.

File: data/generic.json [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+  "config": null,
+  "doc": null,
+  "kmod": null,
+  "permissions": null,
+  "symlinks": null,
+  "upstream": null,
+  "debuginfo": { "ignore": [ "/lib/modules/*",
+			     "/usr/lib/debug/.dwz/*",
+			     "/usr/lib/grub/*" ],
+		 "debuginfo_sections": ".gnu_debugdata" }}
```
[the remark this answers is not in the document]
Comment (frozencemetery) [reply data/generic.json]: I ran it through the emacs linter.  It looks like what you're asking for; lemme know if it's not.

File: data/generic.yaml [hunk 9]
```diff
@@ -650,15 +647,15 @@ pathmigration:
...
     # Any package containing a file in /bin will have a failure
     # reporting that file should be in /usr/bin.
     migrated_paths:
-        - /bin:   /usr/bin
-        - /sbin:  /usr/sbin
-        - /lib:   /usr/lib
-        - /lib64: /usr/lib64
+        /bin:   /usr/bin
+        /sbin:  /usr/sbin
+        /lib:   /usr/lib
```
[the remark this answers is not in the document]
Comment (frozencemetery) [reply data/generic.yaml:657]: Noted.  Eww.

File: include/tyranny.h [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        char *string; /* \0-terminated. */
+        struct y_value **array; /* NULL-terminated. */
+
+        /* keys is NULL-terminated.  values might contain NULLs because YAML
+         * allows you to just not map a dict entry to anything. */
+        struct {
+            char **keys;
+            struct y_value **values;
+        } dict;
+    };
```
[the remark this answers is not in the document]
Comment (frozencemetery) [reply include/tyranny.h]: Thanks, I hate it.

**Comment by frozencemetery** [comment]
Ready for review.  Changes:

- Moved cdson to use COPR for dependency in Fedora, with the plan to add it to
  the distro.  The i686 CI won't pass until then because COPR doesn't provide
  i686 repos because it's 2023.
- Folded in libtyranny and made it C99-compliant.
- Rebased around the new modularity ifdefs and debuginfo_sections changes.
- Fixed many style/formatting issues.
- Backward compat for the array of dicts of strs.
- Explicit compat for badfuncs.forbidden and annocheck.jobs as top-level.
- Add a test for said backward compat.
- Updated dump format to match changes and fixed products.

On the test suite, I want to push back against running it once per config file
type.  Right now on main, a run takes 20m18.303s.  Tripling that and we're
over an hour per run.  I don't think I can in good conscience inflict that on
people.

What I've done instead is add a test that parses all three generic config
files and checks that the rpminspect structures match.  If there're any
specific cases not covered by that you're concerned about, we can add them,
but I think that gets "pretty good" coverage.  Parsing the config file is not
a particular bottleneck for rpminspect - incluiding spinning up the test
machinery, test_cfg.RpminspectCfg.runTest takes 0m0.593s.

This means that any supported config file type can be passed to `rpminspect
-c`.  However, YAML remains the default, in the sense that I haven't modified
any of the detction logic or docs therein.

File: meson.build (hunk 7 above)
[the remark this answers is not in the document]
Comment (dcantrell) [reply meson.build:10]: OK, that's fine.  I'll just leave that to you.

File: data/generic.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dcantrell) [reply data/generic.yaml:602]: Ah, ok, that makes sense.  I don't like YAML's vocabulary for this.

File: include/tyranny.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dcantrell) [reply include/tyranny.h:8]: Reasonable.

**Review by dcantrell — APPROVED** [review]

**Comment by dcantrell** [comment]
> Ready for review. Changes:
> 
>     * Moved cdson to use COPR for dependency in Fedora, with the plan to add it to
>       the distro.  The i686 CI won't pass until then because COPR doesn't provide
>       i686 repos because it's 2023.

Gah.  So I'm considering just removing the i686 builds altogether because I don't know who is using rpminspect on i686-only, but I feel like if I do then someone will complain.

Do you have a package ready for Fedora (and EPEL) for cdson?  I need that part done before merging this because I won't be able to build in Koji from a Copr repo.

>     * Folded in libtyranny and made it C99-compliant.

The 90s thank you.

>     * Rebased around the new modularity ifdefs and debuginfo_sections changes.
> 
>     * Fixed many style/formatting issues.
> 
>     * Backward compat for the array of dicts of strs.
> 
>     * Explicit compat for badfuncs.forbidden and annocheck.jobs as top-level.
> 
>     * Add a test for said backward compat.
> 
>     * Updated dump format to match changes and fixed products.

Reviewed and everything looks good.  Thank you.

> On the test suite, I want to push back against running it once per config file type. Right now on main, a run takes 20m18.303s. Tripling that and we're over an hour per run. I don't think I can in good conscience inflict that on people.

Wow, it actually takes that long?  Damn.

> What I've done instead is add a test that parses all three generic config files and checks that the rpminspect structures match. If there're any specific cases not covered by that you're concerned about, we can add them, but I think that gets "pretty good" coverage. Parsing the config file is not a particular bottleneck for rpminspect - incluiding spinning up the test machinery, test_cfg.RpminspectCfg.runTest takes 0m0.593s.
> 
> This means that any supported config file type can be passed to `rpminspect -c`. However, YAML remains the default, in the sense that I haven't modified any of the detction logic or docs therein.

This is entirely sufficient for test purposes.  I want to ensure that both rpminspect can parse those formats correctly and that when changes are made to the template config files, they are all made in unison.

Only remaining things:

* cdson as a Fedora and EPEL package.  Is that ready to go?
* The libtyranny license.  I was confused if you wanted to keep the MPL one or change it.  Either is fine.

Aside from that, I like what's there.  I can update docs, man pages, other things.  And the packaging stuff of rpminspect-data-generic.

**Comment by frozencemetery** [comment]
> cdson as a Fedora and EPEL package. Is that ready to go?

Thanks; it is now.  Updated deplists accordingly.

(Also updated code to remove gcc extension usage, since rpminspect grew -Werror=pedantic.)

Do you want any changes to default config file detection before merging?  Right now this code still only checks for rpminspect.yaml.

**Comment by dcantrell** [comment]
> Do you want any changes to default config file detection before merging? Right now this code still only checks for rpminspect.yaml.

Let's have that as a separate PR.
